lib/services/plugins-service.ts

```diff
@@ -0,0 +1,190 @@
+		}).future<void>()();
+	}
+	
+	public remove(pluginName: string): IFuture<void> {
+		return (() => {
+			this.executeNpmCommand(PluginsService.UNINSTALL_COMMAND_NAME, pluginName).wait();
+			let showMessage = true;
+			let action = (modulesDestinationPath: string, platform: string, platformData: IPlatformData) => {
+				shelljs.rm("-rf", path.join(modulesDestinationPath, pluginName));
+				this.$logger.out(`Successfully removed plugin ${pluginName} for ${platform}.`);
+				showMessage = false;
+			};
+			this.executeForAllInstalledPlatforms(action);
+			
+			if(showMessage) {
+				this.$logger.out(`Succsessfully removed plugin ${pluginName}`);	
+			}
+		}).future<void>()();
+	}
+	
+	public prepare(pluginData: IPluginData): IFuture<void> {
+		return (() => {
+			let action = (pluginDestinationPath: string, platform: string, platformData: IPlatformData) => {
+				let skipExecution = false;
+				// Process .js files 				
+				let installedFrameworkVersion = this.getInstalledFrameworkVersion(platform).wait();
+				let pluginVersion = (<any>pluginData.platformsData)[platform];
+				if(semver.gt(pluginVersion, installedFrameworkVersion)) {
+					this.$logger.warn(`${pluginData.name} ${pluginVersion} for ${platform} is not compatible with the currently installed framework version ${installedFrameworkVersion}.`);
+					skipExecution = true;  
+				}
+				
+				if(!skipExecution) {
+					this.$fs.ensureDirectoryExists(pluginDestinationPath).wait();
+					shelljs.cp("-R", pluginData.fullPath, pluginDestinationPath);
+				
+					let pluginPlatformsFolderPath = path.join(pluginDestinationPath, pluginData.name, "platforms", platform);
+					let pluginConfigurationFilePath = path.join(pluginPlatformsFolderPath, platformData.configurationFileName);					
+					if(this.$fs.exists(pluginConfigurationFilePath).wait()) {
+						let pluginConfigurationFileContent = this.$fs.readFile(pluginConfigurationFilePath).wait().toString();
+						let configurationFileContent = this.$fs.readFile(platformData.configurationFilePath).wait().toString();
+						let resultXml = this.mergeXml(pluginConfigurationFileContent, configurationFileContent).wait();
+						this.$fs.writeFile(platformData.configurationFilePath, resultXml).wait();	
+					}
+					
+					if(this.$fs.exists(pluginPlatformsFolderPath).wait()) {
+						shelljs.rm("-rf", pluginPlatformsFolderPath);			
+					}						
+					
+					// TODO: Add libraries
+					
+					// Show message
+					this.$logger.out(`Successfully prepared plugin ${pluginData.name} for ${platform}.`);
+				}
+			};
+			
+			this.executeForAllInstalledPlatforms(action);
+		}).future<void>()();
+	}
+	
+	public getAllInstalledPlugins(): IFuture<IPluginData[]> {
+		return (() => {
+			let nodeModules = this.$fs.readDirectory(this.nodeModulesPath).wait();
+			let plugins: IPluginData[] = [];
+			_.each(nodeModules, nodeModuleName => {
+				var nodeModuleData = this.getNodeModuleData(nodeModuleName);
```

Comment:
also consider using filter + map:
```TypeScript
return _(nodeModules)
          .map(nm => this.getNodeModuleData(nm))
          .filter(nm => nm.isPlugin)
          .map(nm => this.convertToPluginData(nodeModuleData))
          .value();
```
After writing this I'm not sure my suggestion is correct. Maybe your code is even faster than this chain...